SAMZA-1225: Add demo examples for programming with the fluent API

vjagadish1989 opened the pull request:
This allows new users to try these samples out, rapidly experiment with them, and view results.

The following are the examples added:

1. A `PageView` sessionizer job
2. A joiner job that joins a stream of `PageView`s and `AdClick`s
3. A windowed counter that groups `PageView`s by some key, and emits output
4. A simple `PageView` filter that demonstrates stateless transforms

jmakes commented on src/main/java/samza/examples/cookbook/PageViewAdClickJoiner.java:
```diff
@@ -0,0 +1,115 @@
+    Function<String, String> adClickKeyFn = adClick -> new AdClick(adClick).getPageId();
+
+    MessageStream<String> pageViewRepartitioned = pageViews.partitionBy(pageViewKeyFn);
+    MessageStream<String> adClickRepartitioned = adClicks.partitionBy(adClickKeyFn);
```
I prefer inlining the key functions here, unless the point is to demonstrate their types.

prateekm replied:
+1, same for the FilterFunction in one of the other examples. It's fairly uncommon to need to extract/reuse lambdas in fluent APIs I've worked with. Also, the operator chain at the bottom is more descriptive with them inlined.

The JoinFunction OTOH should be extracted, maybe as an inner class?

vjagadish1989 replied:
+1. To add, I prefer doing this in cases except in cases where `keyFn` is used multiple times. For instance, it's common to use the same `keyFn` for the `partitionBy` and the `window` operators.

prateekm commented on gradle.properties:
```diff
@@ -18,7 +18,7 @@
 
 SAMZA_VERSION=0.13.0-SNAPSHOT
-KAFKA_VERSION=0.10.0.1
+KAFKA_VERSION=0.10.1.1
```
Thanks for fixing these.

prateekm commented on src/main/config/pageview-adclick-joiner.properties:
```diff
@@ -0,0 +1,46 @@
+systems.kafka.samza.factory=org.apache.samza.system.kafka.KafkaSystemFactory
+systems.kafka.samza.msg.serde=string
+systems.kafka.samza.key.serde=string
+systems.kafka.consumer.zookeeper.connect=localhost:2181/
```
Is the trailing / required?

vjagadish1989 replied:
Probably not, all our hello world jobs have it. I'll verify it, and remove it if it's not needed.

prateekm commented on src/main/config/pageview-adclick-joiner.properties:
```diff
@@ -0,0 +1,46 @@
+task.window.ms=2000
+
+# Serializers
+serializers.registry.json.class=org.apache.samza.serializers.JsonSerdeFactory
```
Looks like this is unused in all of these. Maybe remove?

vjagadish1989 replied:
It's useful as a helper should you need a `Json` serde for your inputs/stores.

prateekm commented on src/main/config/tumbling-pageview-counter.properties:
```diff
@@ -0,0 +1,46 @@
+
+# Task
+app.class=samza.examples.cookbook.TumblingPageViewCounterApp
+task.inputs=kafka.pageviews
```
Very Minor: Maybe tumbling-pageview-counter-input?

prateekm commented on src/main/java/samza/examples/cookbook/AdClick.java:
```diff
@@ -0,0 +1,59 @@
+  private final String pageId;
+
+  public AdClick(String message) {
+
```
Minor: Delete extra newline.

prateekm commented on src/main/java/samza/examples/cookbook/AdClick.java:
```diff
@@ -0,0 +1,59 @@
+  public String getPageId() {
+    return pageId;
+  }
+
```
Minor: Delete extra newline.

prateekm commented on src/main/java/samza/examples/cookbook/TumblingPageViewCounterApp.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.function.Function;
+
+/**
+ * In this example, we group a stream of page views by country, and compute the number of page views over a tumbling time
```
Minor: Maybe explain what tumbling windows are.

prateekm commented on src/main/java/samza/examples/cookbook/TumblingPageViewCounterApp.java:
```diff
@@ -0,0 +1,87 @@
+
+    MessageStream<String> pageViews = graph.<String, String, String>getInputStream(INPUT_TOPIC, (k, v) -> v);
+
+    OutputStream<String, String, WindowPane<String, Collection<String>>> outputStream = graph
```
Minor: IMHO it'll be a little cleaner to add a .map after the window to do the counting and send that value. Or maybe use an aggregator here? 

As an aside, it's a little confusing that the WindowPane has getKey and getMessage. Maybe we should rename it to something like getWindowKey and getWindowContents (or s/Window/Pane)?

vjagadish1989 replied:
I think we have been having prior discussions on `WindowPane`.  I'm +1 for `getWindowKey` and `getWindowContents`. 

It also ensures we don't have `WindowPane.getKey.getKey()` which IMO is a slightly awkward usage.

Do you mind tracking this as a part of a clean-up/ renaming patch we do?

prateekm commented on src/main/java/samza/examples/cookbook/PageViewSessionizerApp.java:
```diff
@@ -0,0 +1,87 @@
+
+/**
+ * In this example, we group page views by userId into sessions, and compute the number of page views for each user
+ * session. A session is considered closed when there is no user activity for a 3 second duration.
```
Very minor: Maybe increase to 5 minutes or something. 3 second sessions seem very short.

vjagadish1989 replied:
I prefer a shorter session interval (need not be a 3 sec one) so that users don't have to wait for 5 minutes to observe a meaningful output (since this is a toy use-case). 

They can always fiddle around with the session interval, and observe how results change.

vjagadish1989 commented:
Address @prateekm and @jmakes review feedback.

vjagadish1989 commented:
Address @prateekm's feedback:
 
- Added example that uses a `foldLeftFunction` instead.
 - Changed input topic name to `pageview-tumbling-input` 
 - Changed output topic name to `pageview-tumbling-output` 
 - Verified that all jobs, and configs actually run with these changes.
